categorize_notify/program.py

```diff
@@ -9,64 +9,106 @@
 """
 
 import base64
+from datetime import datetime, timezone
+import os
 import time
 
 import autokitteh
 from autokitteh import google, openai, slack
 
 
+POLL_INTERVAL = os.getenv("POLL_INTERVAL")
 SLACK_CHANNELS = ["demos", "engineering", "ui"]
+gmail = google.gmail_client("my_gmail").users()
+processed_message_ids = set()
+start_time = datetime.now(timezone.utc).timestamp()
 
 
 def on_http_get(event):
-    total_messages = None
     while True:
-        total_messages = _poll_inbox(total_messages)
-        time.sleep(10)
+        _poll_inbox()
+        time.sleep(float(POLL_INTERVAL))
+
+
+def _poll_inbox():
+    token = None
+    current_message_ids = set()
+
+    while True:
+        results = get_new_inbox_messages(token)
+
+        current_message_ids.update({msg["id"] for msg in results.get("messages", [])})
+
+        token = results.get("nextPageToken")
+        if not token:
+            break
+
+    new_message_ids = current_message_ids - processed_message_ids
+
+    for message_id in new_message_ids:
+        _process_email(message_id, start_time)
+
+    processed_message_ids.update(new_message_ids)
 
 
 @autokitteh.activity
-def _poll_inbox(prev_total: int):
-    gmail = google.gmail_client("my_gmail").users()
-    curr_total = gmail.getProfile(userId="me").execute()["messagesTotal"]
-    # Note: This is not meant to be a robust solution for handling email operations.
-    if prev_total and curr_total > prev_total:
-        new_email_count = curr_total - prev_total
-        message_ids = _get_latest_message_ids(gmail, new_email_count)
-        for message_id in message_ids:
-            _process_email(gmail, message_id)
+def get_new_inbox_messages(token):
+    query = "in:inbox -in:drafts"
+    if token:
+        gmail_list = gmail.messages().list(userId="me", q=query, pageToken=token)
+        results = gmail_list.execute()
+    else:
+        results = gmail.messages().list(userId="me", q=query).execute()
```

Comment:
Please test if passing the tokens `""` and `None` has the same effect like not specifying a token at all - if so, remove the `if` condition and the `else` branch.

Reply:
I thought about that too and I don't understand why but when I do it the way you're suggesting I get this error:
[error.txt](https://github.com/user-attachments/files/16593270/error.txt)

I ran it twice each way and got consistent results. Which I understand doesn't prove anything. I'll try a few more times later.

Reply:
After talking with Itay, I decided to remove pagination entirely. I get the first 100 results by default and they seem to be in reverse chronological order although I could not find any guarantees from Google.

I think this is still better than paginating the entire inbox.